To configure multiple public network using ceph config option

sunilangadi2 opened the pull request:
Signed-off-by: Sunil Angadi <sangadi@redhat.com>

# Description
Test case: CEPH-83575132
Fixes: RHCEPHQE-4071
Added new test suite to deploy ceph cluster with multiple public network using cephadm.
also validated configured public network using `config get` option.
Test result:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-BH4Q3G/

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

vamahaja commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+    mask = ceph_pub_network.rsplit(sep="/")
+    # logic to create new public ip by incrementing mask count
+    new_mask = int(mask[1]) + 1
+    new_pub_network = mask[0] + "/" + str(new_mask)
```
I would suggest to use ```f"{mask[0]}/{str(new_mask)}"```

vamahaja commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+    # logic to create new public ip by incrementing mask count
+    new_mask = int(mask[1]) + 1
+    new_pub_network = mask[0] + "/" + str(new_mask)
+    multiple_network = f"{ceph_pub_network}" + "," + f"{new_pub_network}"
```
Below code shoudl work 
```
f"{ceph_pub_network},{new_pub_network}"
```

vamahaja commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+    new_mask = int(mask[1]) + 1
+    new_pub_network = mask[0] + "/" + str(new_mask)
+    multiple_network = f"{ceph_pub_network}" + "," + f"{new_pub_network}"
+    cmd = ["ceph", "config", "set", "mon", "public_network", f"{multiple_network}"]
```
instead `f"{multiple_network}"`, `multiple_network` can be used

vamahaja commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+    new_pub_network = mask[0] + "/" + str(new_mask)
+    multiple_network = f"{ceph_pub_network}" + "," + f"{new_pub_network}"
+    cmd = ["ceph", "config", "set", "mon", "public_network", f"{multiple_network}"]
+    try:
+        out, err = cls.shell(args=cmd)
+        log.debug("Set multiple public network to ceph mon successfully")
+    except Exception:
+        log.error("Failed to set multiple public network to ceph")
+        log.error(traceback.format_exc())
+        return False
```
Please avoid using general exception. Use exact exception which will be raised by calling function and comprare expected error messages in except if possible

vamahaja commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+    public_network, _ = cls.shell(
+        args=["ceph", "config", "get", "mon", "public_network"]
+    )
+    if public_network.rstrip() == multiple_network:
+        return True
+    else:
+        return False
```
Below code works same way -
```
    if public_network.rstrip() == multiple_network:
        return True

    return False
```

vamahaja commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+
+def run(ceph_cluster, **kw):
+    """
+    To configure any specific ceph commands using ceph config option.
```
IMO below description suites more here -
```
Test to validate to configure any specific ceph commands using ceph config option.
```

vamahaja commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+    config = kw.get("config")
+    instance = CephAdmin(cluster=ceph_cluster, **config)
+    log.info("adding new public network to ceph")
+    if set_multiple_public_network(instance):
+        log.info("Verified multiple public network to ceph")
+    else:
+        log.error("Test failed to set multiple public network to ceph")
+        return 1
+    return 0
```
Below code will work in same way 
```
   if set_multiple_public_network(instance):
        log.info("Verified multiple public network to ceph")
        return 0
   
   log.error("Test failed to set multiple public network to ceph")
   return 1
```

manasagowri commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+    """
+    out, _ = cls.shell(args=["ceph", "config", "get", "mon", "public_network"])
+    ceph_pub_network = out.rstrip()
+    mask = ceph_pub_network.rsplit(sep="/")
```
This step might not work if already multiple public networks have been configured, and user wants to add another to the list. If possible please test that scenario as well.

sunilangadi2 replied:
yes manasa i handled this situation now we can set one more network to existing multiple network
```
2022-08-22 03:18:43,326 - INFO - cephci.test_config:110 - Running command cephadm -v shell -- ceph config set mon public_network 10.0.208.0/22,10.0.208.0/23,10.0.208.0/24,10.0.208.0/25,10.0.208.0/26 on 10.0.210.235 timeout 600
2022-08-22 03:18:45,076 - INFO - cephci.test_config:110 - Command completed successfully
```

manasagowri commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,60 @@
+        cls: cephadm instance object
+
+    Returns:
+        False -> Fail, True -> Pass
```
Could you please add more description as to, when you say fail what failed (setting public network failed or any other validation post that)? similarly for pass.

sunilangadi2 commented:
Thanks, @vamahaja @manasagowri for the review,
Addressed your review comments please review it.
Test result: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-KHWQEM

sunilkumarn417 commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,62 @@
+    # logic to create new public ip by incrementing mask count
+    new_mask = int(mask[-1]) + 1
+    new_pub_network = f"{mask[0]}/{str(new_mask)}"
+    multiple_network = f"{ceph_pub_network},{new_pub_network}"
```
@sunilangadi2 can you please point me on how we are getting multiple networks?

I don't think `ceph config get mon public_network` will give us multiple networks.

sunilangadi2 replied:
@sunilkumarn417 Yes from cephci side by deafult bootstrap sets one public network 
But initially, there is bug with `cephadm` which is not able to set multiple network using `ceph config set mon public_network` so to test that scenario  created one more public network by incrementing mask count which is not really public network IP ,
We are just validating cephadm CLI functionality not the whole network functionality.

sunilkumarn417 commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,62 @@
+    return False
+
+
+def run(ceph_cluster, **kw):
```
To test this case, I think we need ceph nodes to be in different networks.
- Please check is there any possiblity in PSI. because In IBM we have only one network in VPC.
- Please check if you want this to be executed in IBM test pipeline or could cover this in PSI.

sunilangadi2 replied:
This test for cephadm-specific CLI functionality not the multiple public network functionality.
Hence no need of real different public network infra.

sunilkumarn417 commented:
@pdhiran @psathyan

sunilkumarn417 commented:
As per discussion on validation, let's verify the public network configuration post bootstrap without explicitly applying `ceph config set ` command and with acutal networks associated with the nodes. 

Thanks @pdhiran @rahullepakshi .

sunilangadi2 commented:
> As per discussion on validation, let's verify the public network configuration post bootstrap without explicitly applying `ceph config set ` command and with acutal networks associated with the nodes.
> 
> Thanks @pdhiran @rahullepakshi .

Updated the PR as per suggestion, Please review it.
Test results [Passed] : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-QLFWCQ

sunilkumarn417 commented on suites/pacific/cephadm/tier-1_multiple_subnet.yaml:
```diff
@@ -0,0 +1,229 @@
+                  nodes:
+                    - node5
+      abort-on-fail: true
+  - test:
+      name: Apply Monitor service
+      desc: Apply monitor service using label mon
+      module: test_mon.py
+      polarion-id: CEPH-83573772
+      config:
+        command: apply
+        service: mon
+        base_cmd_args:
+          verbose: true
+        args:
+          placement:
+            label: mon
+      destroy-cluster: false
+      abort-on-fail: true
+  - test:
+      name: Apply Manager service
+      desc: Apply manager service using label mgr
+      module: test_mgr.py
+      polarion-id: CEPH-83573777
+      config:
+        command: apply
+        service: mgr
+        base_cmd_args:
+          verbose: true
+        args:
+          placement:
+            label: mgr
+      destroy-cluster: false
+      abort-on-fail: true
```
can we combine these test cases, since we are testing only multiple networks.

sunilkumarn417 commented on suites/pacific/cephadm/tier-1_multiple_subnet.yaml:
```diff
@@ -0,0 +1,229 @@
+          all-available-devices: true
+      destroy-cluster: false
+      abort-on-fail: true
+  - test:
```
do we need thsi test case

sunilangadi2 replied:
so far now i think we don't have any test suite which using multiple public network configured cluster,
Hence I though it's good to add other service as well.
In order to check other service deployments with different subnet cluster.

sunilkumarn417 commented on suites/pacific/cephadm/tier-1_multiple_subnet.yaml:
```diff
@@ -0,0 +1,229 @@
+                    label: mds
+      destroy-cluster: false
+      abort-on-fail: true
+  - test:
```
do we need object?

sunilangadi2 replied:
same answer to cover other service in multiple public network configured cluster,
Hence added RGW as well

sunilkumarn417 commented on suites/pacific/cephadm/tier-1_multiple_subnet.yaml:
```diff
@@ -0,0 +1,229 @@
+        check_status: True
+        commands:
+          - "ceph status"
+  - test:
```
do we need this test case?

sunilangadi2 replied:
yes as @pdhiran suggested to check mon quorum status after bootstrap with multiple public network cluster,
Hence added it.

sunilangadi2 commented on suites/pacific/cephadm/tier-1_multiple_subnet.yaml:
```diff
@@ -0,0 +1,229 @@
+      polarion-id: CEPH-83573627
+      module: test_election_strategies.py
+      desc: Change Mon election strategies and verify status
+  - test:
```
This test is for changing the election strategies of mon and observes mon quorum behavior,
So it needs mon daemons to be deployed, it's good to test this with services hence added it over here.

sunilangadi2 commented on tests/cephadm/test_config.py:
```diff
@@ -0,0 +1,42 @@
+    config = kw["config"]
+    instance = CephAdmin(cluster=ceph_cluster, **config)
+    configured_network = config["public_network"]
+    if verify_multiple_public_network(instance, configured_network):
```
Nothing also even as per the BZ we just have to check config option got set properly in `public network` option.

manasagowri reviewed (approved):
LGTM

sunilangadi2 commented:
@sunilkumarn417 @manasagowri As discussed

- Removed the tests which is not needed for this bz automation,
- added new folder called `customer_bz` in cephadm suite
- updated suite and conf name with OpenStack only as it needs to run only in OpenStack
- updated doc string in suite and conf file.

Test Results - [PASSED]
logs: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-BH4Q3G/

Thanks for the comments, please review it.

sunilkumarn417 reviewed (commented):
Hi @sunilangadi2 ,

LGTM.

Can you please add this test suite in release metadata.yaml and please tag psi tag to get it execute it only on psi.
